homeassistant/components/homekit_controller/climate.py

```diff
@@ -170,8 +182,57 @@ def get_characteristic_types(self) -> list[str]:
             CharacteristicsTypes.TEMPERATURE_COOLING_THRESHOLD,
             CharacteristicsTypes.TEMPERATURE_HEATING_THRESHOLD,
             CharacteristicsTypes.SWING_MODE,
+            CharacteristicsTypes.ROTATION_SPEED,
         ]
 
+    def _get_rotation_speed_range(self) -> tuple[float, float]:
+        return (
+            round(self.service[CharacteristicsTypes.ROTATION_SPEED].minValue or 0) + 1,
+            round(self.service[CharacteristicsTypes.ROTATION_SPEED].maxValue or 100),
+        )
+
+    @property
+    def fan_modes(self) -> list[str] | None:
+        """Return the available fan modes."""
+        if self.service.has(CharacteristicsTypes.ROTATION_SPEED):
+            return [FAN_OFF, FAN_LOW, FAN_MEDIUM, FAN_HIGH]
+        return None
```

Comment:
According to codecov, this line is uncovered by tests.

I don't think it is possible for it to be covered, because it's only called when the FAN_MODE feature is present:

https://github.com/home-assistant/core/blob/4e80fcc6f1f279d553864dd5dbfc30df6e0a31bb/homeassistant/components/climate/__init__.py#L279-L280

The situation is the same with swing mode - we only check self.swing_modes if the feature is present. Based on what `swing_modes` and others like it do, I think the right thing to do to make codecov happy is just:

```
    @property
    def fan_modes(self) -> list[str]:
        """Return the available fan modes."""
        return [FAN_OFF, FAN_LOW, FAN_MEDIUM, FAN_HIGH]
```

But I would like a 2nd opinion from @bdraco there I think.

Reply:
indeed i copied from swing :)

Reply:
let me know @bdraco and @Jc2k what do u think is the best solution.
M.

Reply:
https://github.com/home-assistant/core/pull/98291/files#r1299472291

Reply:
committed the suggestion.

Reply:
You need to update the type signature as well now it no longer returns None. See mine above ^

Reply:
u are right ,fixed.